lua/lspconfig/util.lua

```diff
@@ -453,38 +446,45 @@ function M.root_pattern(...)
   end
   return function(startpath)
     startpath = M.strip_archive_subpath(startpath)
-    return M.search_ancestors(startpath, matcher)
+    return function()
+      local guard = 100
+      for path in M.path.iterate_parents(startpath) do
+        -- Prevent infinite recursion if our algorithm breaks
+        guard = guard - 1
+        if guard == 0 then
+          return
+        end
+
+        if matcher(path) then
+          return path
+        end
+      end
+    end
   end
 end
+
 function M.find_git_ancestor(startpath)
-  return M.search_ancestors(startpath, function(path)
-    -- Support git directories and git files (worktrees)
-    if M.path.is_dir(M.path.join(path, '.git')) or M.path.is_file(M.path.join(path, '.git')) then
-      return path
-    end
-  end)
+  return function()
+    return searcher(startpath, { '.git' })
+  end
 end
+
 function M.find_mercurial_ancestor(startpath)
```

Comment:
Maybe also add a `tail` param here and to the other helpers that defaults to `false`? This would not break bc.

Reply:
yep . just for test i set it to true. just search on github. found lots of function in utils are used in people config ... can't do break change ..

Reply:
Needs more consideration and design here. take compatibility issues into consideration.  🧐